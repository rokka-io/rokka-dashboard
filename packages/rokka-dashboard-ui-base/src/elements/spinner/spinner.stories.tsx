import { boolean, color } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import { colors } from '../../identity/colors/colors';

import { Spinner } from './spinner';

const ExampleContainer = styled.div`
  position: relative;
  background: ${colors.brand.primary};
  width: 200px;
  height: 200px;
`;

storiesOf('Elements / ActivityIndicators', module).add('Spinner', () => (
  <ExampleContainer>
    <Spinner small={boolean('small', false)} backgroundColor={color('background color', colors.tints.black)} />
  </ExampleContainer>
));
