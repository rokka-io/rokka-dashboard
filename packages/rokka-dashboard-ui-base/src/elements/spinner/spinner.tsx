import React from 'react';
import styled, { keyframes } from 'styled-components';
import { colors } from '../../identity/colors/colors';

interface SpinnerProps {
  small?: boolean;
  backgroundColor?: string;
}

export function Spinner({ small, backgroundColor }: SpinnerProps) {
  return (
    <FoldingCube small={small}>
      <Cube backgroundColor={backgroundColor} />
      <Cube2 backgroundColor={backgroundColor} />
      <Cube4 backgroundColor={backgroundColor} />
      <Cube3 backgroundColor={backgroundColor} />
    </FoldingCube>
  );
}

interface FoldingCubeProps {
  small?: boolean;
}

// TODO: rem instead of px?
const FoldingCube = styled.div<FoldingCubeProps>`
  margin: ${({ small = false }) => (small ? 0 : '20px auto')};
  width: ${({ small = false }) => (small ? '16px' : '40px')};
  height: ${({ small = false }) => (small ? '16px' : '40px')};
  position: relative;
  transform: rotateZ(45deg);
`;

const foldCubeAngle = keyframes`
  0%, 10% {
    transform: perspective(140px) rotateX(-180deg);
    opacity: 0;
  }
  25%, 75% {
    transform: perspective(140px) rotateX(0deg);
    opacity: 1;
  }
  90%, 100% {
    transform: perspective(140px) rotateY(180deg);
    opacity: 0;
  }
`;

interface CubeProps {
  backgroundColor?: string;
}

const Cube = styled.div<CubeProps>`
  float: left;
  width: 50%;
  height: 50%;
  position: relative;
  transform: scale(1.1);

  :before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: ${({ backgroundColor = colors.tints.black }) => backgroundColor};
    animation: ${foldCubeAngle} 2.4s infinite linear both;
    transform-origin: 100% 100%;
  }
`;

const Cube2 = styled(Cube)`
  transform: scale(1.1) rotateZ(90deg);

  :before {
    animation-delay: 0.3s;
  }
`;
const Cube3 = styled(Cube)`
  transform: scale(1.1) rotateZ(180deg);

  :before {
    animation-delay: 0.6s;
  }
`;
const Cube4 = styled(Cube)`
  transform: scale(1.1) rotateZ(270deg);

  :before {
    animation-delay: 0.9s;
  }
`;
