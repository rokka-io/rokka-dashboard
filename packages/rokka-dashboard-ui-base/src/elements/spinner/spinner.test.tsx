import { shallow } from 'enzyme';
import React from 'react';

import { Spinner } from './spinner';

describe('Elements', () => {
  describe('ActivityIndicators', () => {
    it('renders a spinner correctly.', () => {
      const component = shallow(<Spinner />);

      expect(component).toMatchSnapshot();
    });
    it('renders a small spinner correctly.', () => {
      const component = shallow(<Spinner small={true} />);

      expect(component).toMatchSnapshot();
    });
    it('renders a white spinner correctly.', () => {
      const component = shallow(<Spinner backgroundColor="#fff" />);

      expect(component).toMatchSnapshot();
    });
  });
});
