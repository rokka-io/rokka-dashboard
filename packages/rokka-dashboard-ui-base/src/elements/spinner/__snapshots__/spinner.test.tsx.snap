// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements ActivityIndicators renders a small spinner correctly. 1`] = `
<styled.div
  small={true}
>
  <styled.div />
  <Styled(styled.div) />
  <Styled(styled.div) />
  <Styled(styled.div) />
</styled.div>
`;

exports[`Elements ActivityIndicators renders a spinner correctly. 1`] = `
<styled.div>
  <styled.div />
  <Styled(styled.div) />
  <Styled(styled.div) />
  <Styled(styled.div) />
</styled.div>
`;

exports[`Elements ActivityIndicators renders a white spinner correctly. 1`] = `
<styled.div>
  <styled.div
    backgroundColor="#fff"
  />
  <Styled(styled.div)
    backgroundColor="#fff"
  />
  <Styled(styled.div)
    backgroundColor="#fff"
  />
  <Styled(styled.div)
    backgroundColor="#fff"
  />
</styled.div>
`;
